## <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:150%; text-align:center; border-radius:10px 10px;">Çalışmanızı daha verimli hale getirecek 8 Python Pandas value_counts() püf noktası</p>

 - Herhangi bir veri projesine başlamadan önce bir adım geriye gitmeli ve veri kümesiyle herhangi bir şey yapmadan önce ona bakmalısınız. Keşifsel Veri Analizi (EDA) veri analizinin herhangi bir parçası kadar önemlidir, çünkü gerçek veri kümeleri çok karmaşıktır ve verilerinizi bilmiyorsanız birçok şey ters gidebilir.

 - Pandas kütüphanesi bu amaç için çeşitli kullanışlı işlevlerle donatılmıştır ve ``"value_counts()"`` bunlardan biridir. Pandas, bir pandas veri çerçevesindeki benzersiz değer sayılarını sıralanmış düzende ``"value_counts()"`` içeren bir nesneyi döndürür. Ancak çoğu kullanıcı bu işlevin yalnızca varsayılan parametrelerle kullanılamayacağını gözden kaçırma eğilimindedir. Bu makalede, varsayılan parametreleri değiştirerek Pandasda nasıl daha fazla değer elde edebileceğinizi ve size zaman kazandıracak birkaç ek hileyi göstereceğim.
 
<img src=https://re-thought.com/content/images/size/w2000/2020/05/zac-ong-hPOFScEaZcA-unsplash.jpg width="300" height="200">


**🧐 [Makaleye buradan ulasabilirsiniz](https://re-thought.com/pandas-value_counts/)**

s<dgsg

## value_counts() işlevi nedir?

İşlev, benzersiz değerlerin sayısını içeren bir Seri ``"value_counts()"`` elde etmek için kullanılır. Ortaya çıkan nesne, ilk öğenin en sık görülen öğe olması için azalan sırada olacaktır. NA değerlerini varsayılan olarak hariç tutar.

### Syntax

``"df['your_column'].value_counts()"`` Bu belirtilen sütundaki benzersiz oluşumların (unique değerlerin) sayısını döndürecektir.
``"value_counts()"`` Dataframe lerde değil, yalnızca Seri lerde çalıştığını unutmamak önemlidir.

Sonuç olarak, yalnızca bir parantez ``"df['your_column']"`` içerir, iki parantez ``"df[['your_column']]"`` içermez.


## Parameters
 - **normalize (bool, default False)** - True ise döndürülen nesne benzersiz değerlerin göreceli frekanslarını içerecektir.
 - **sort (bool, default True)** - Frekanslara göre sıralayın.
 - **ascending (bool, default False)** - Artan düzende sıralayın.
 - **bins (int, optional)** - Değerleri saymak yerine, bunları yarı açık kutular halinde gruplandırın, ``"pd.cut"`` bu kolaylık yalnızca sayısal verilerle çalışır.
 - **dropna (bool, default True)** - NaN sayılarını dahil etmeyin

In [2]:
import pandas as pd

In [7]:
df = pd.read_csv('coursea_data.csv')

In [8]:
df.head(10)

Unnamed: 0                                       course_title  \
0         134  (ISC)² Systems Security Certified Practitioner...   
1         743  A Crash Course in Causality:  Inferring Causal...   
2         874                     A Crash Course in Data Science   
3         413                            A Law Student's Toolkit   
4         635                A Life of Happiness and Fulfillment   
5         661  ADHD: Everyday Strategies for Elementary Students   
6          54                                    AI For Everyone   
7         488                           AI For Medical Treatment   
8          58                        AI Foundations for Everyone   
9         129                           AI for Medical Diagnosis   

          course_organization course_Certificate_type  course_rating  \
0                      (ISC)²          SPECIALIZATION            4.7   
1  University of Pennsylvania                  COURSE            4.7   
2    Johns Hopkins University                  COURSE            4.5   
3             Yale University                  COURSE            4.7   
4   Indian School of Business                  COURSE            4.8   
5       University at Buffalo                  COURSE            4.7   
6             deeplearning.ai                  COURSE            4.8   
7             deeplearning.ai                  COURSE            4.8   
8                         IBM          SPECIALIZATION            4.7   
9             deeplearning.ai                  COURSE            4.7   

  course_difficulty course_students_enrolled  
0          Beginner                     5.3k  
1      Intermediate                      17k  
2             Mixed                     130k  
3             Mixed                      91k  
4             Mixed                     320k  
5          Beginner                      39k  
6          Beginner                     350k  
7      Intermediate                     2.4k  
8          Beginner                      61k  
9      Intermediate                      12k

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 7 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                891 non-null    int64  
 1   course_title              891 non-null    object 
 2   course_organization       891 non-null    object 
 3   course_Certificate_type   891 non-null    object 
 4   course_rating             891 non-null    float64
 5   course_difficulty         891 non-null    object 
 6   course_students_enrolled  891 non-null    object 
dtypes: float64(1), int64(1), object(5)
memory usage: 48.9+ KB


#### Bu bize veri setimizde 891 kaydımızın olduğunu ve herhangi bir NaN değerimizin olmadığını söyler.

## 1. value_counts() with default parameters

Veri çerçevemizden "course_difficulty" sütununa ilişkin sayıları alacağız.

In [13]:
# course_difficulty sütunu için tüm unique değerlerin sayısı

df['course_difficulty'].value_counts()

course_difficulty
Beginner        487
Intermediate    198
Mixed           187
Advanced         19
Name: count, dtype: int64

Bu fonksiyon ``"value_counts()"`` verilen dizindeki tüm benzersiz(unique) değerlerin sayısını herhangi bir boş değer olmadan azalan sırada döndürür. Maksimum kursların Beginner seviyesinde olduğunu, ardından sırası ile Intermediate, Mixed ve Advanced seviyesinde olduğunu hızlı bir şekilde görebiliriz.

## 2. value_counts() in ascending order
 ``"value_counts()"`` Tarafından döndürülen seriler default olarak azalan sıradadır. ``"ascending"`` Parametreyi ``"True"`` olarak ayarlayarak durumu tersine çevirebiliriz.

In [14]:
# course_difficulty sütunu için tüm unique değerlerin artan sırada sayısı

df['course_difficulty'].value_counts(ascending=True)

course_difficulty
Advanced         19
Mixed           187
Intermediate    198
Beginner        487
Name: count, dtype: int64

## 3. value_counts() sorted alphabetically

Bazı durumlarda ``"value_counts()"``ile döndürülenleri alfabetik sırada görüntülememiz gerekebilir.

Bu ``"value_counts()"``'tan sonra ``"sort_index(ascending=True)"``sıralama indeksi eklenerek kolayca yapılabilir.

In [16]:
# value_counts() alfabetik olarak sıralanır.
df['course_difficulty'].value_counts().sort_index(ascending=True)

course_difficulty
Advanced         19
Beginner        487
Intermediate    198
Mixed           187
Name: count, dtype: int64

In [17]:
# value_counts() ters alfabetik olarak sıralanır.
df['course_difficulty'].value_counts().sort_index(ascending=False)

course_difficulty
Mixed           187
Intermediate    198
Beginner        487
Advanced         19
Name: count, dtype: int64

## 4. Pandas value_counts(): sort by value, then alphabetically


In [23]:
# Bu örnek için biraz farklı bir veri çerçevesi kullanalım
df_fruit = pd.DataFrame({
    'fruit':
        ['sharon fruit']*5 + ['apples']*5 + ['bananas']*3 + 
        ['nectarines']*3 + ['carrots']*3 + ['apricots'] + ['mango']*2})

In [24]:
df_fruit

fruit
0   sharon fruit
1   sharon fruit
2   sharon fruit
3   sharon fruit
4   sharon fruit
5         apples
6         apples
7         apples
8         apples
9         apples
10       bananas
11       bananas
12       bananas
13    nectarines
14    nectarines
15    nectarines
16       carrots
17       carrots
18       carrots
19      apricots
20         mango
21         mango

In [ ]:
 sort_index(ascending=False)Bu, value_counts() işlevini ve ile birleştirerek yapılabilir sort_values(ascending=False).

In [25]:
# Burada çıktının önce değer sayılarına göre, sonra da meyve adına göre alfabetik olarak sıralanmasını istiyoruz.

df_fruit['fruit'].value_counts().sort_index(ascending=False).sort_values(ascending=False)

fruit
sharon fruit    5
apples          5
nectarines      3
carrots         3
bananas         3
mango           2
apricots        1
Name: count, dtype: int64

###### Önce değer sayılarına göre ``"sort_index(ascending=False)"`` sonra da meyve adına göre alfabetik ``"sort_values(ascending=False)"``

## 5. value_counts() persentage counts or relative frequencies of the unique values
Bazen % almak normal sayıdan daha iyidir. ``"normalize=True"`` ayarlandığında döndürülen nesne benzersiz(unique) değerlerin frekanslarını içerir.

Default olarak ``"normalize=False"`` dur.

In [26]:
# value_counts percentage view
df['course_difficulty'].value_counts(normalize=True)

course_difficulty
Beginner        0.546577
Intermediate    0.222222
Mixed           0.209877
Advanced        0.021324
Name: proportion, dtype: float64

## 6. value_counts() to bin continuous data into discrete intervals
Bu genellikle yeterince kullanılmayan harika bir hile.  ``"bin"``Parametre yardımıyla sürekli verileri ayrık aralıklara bölmek için kullanılabilir. Bu seçenek yalnızca sayısal verilerle çalışır. ``"pd.cut."``Fonksiyonuna benzer.

In [27]:
# applying value_counts with default parameters 
df['course_rating'].value_counts()

course_rating
4.8    256
4.7    251
4.6    168
4.5     80
4.9     68
4.4     34
4.3     15
4.2     10
5.0      2
4.0      2
3.9      2
3.3      1
4.1      1
3.7      1
Name: count, dtype: int64

In [28]:
# Sütunun sayımlarını 4 bölmeye gruplayalım.
# applying value_counts on a numerical column 
# with the bin parameter
df['course_rating'].value_counts(bins=4)

course_rating
(4.575, 5.0]      745
(4.15, 4.575]     139
(3.725, 4.15]       5
(3.297, 3.725]      2
Name: count, dtype: int64

######  ``"bin"``, iletilen fikrin anlaşılmasını kolaylaştırır. Toplam nüfus içinde çoğu kişinin derslere 4,5'un üzerinde puan verdiğini rahatlıkla görebiliriz. Derecelendirmenin 4,15'in altında olduğu yalnızca birkaç aykırı değer dışında (4,15'ten düşük derecelendirilmiş yalnızca 7 kurs).

## 7. value_counts() displaying the NaN values
Varsayılan olarak boş değerlerin sayısı sonuçtan hariç tutulur.  Ancak, ``"dropna"``parametreyi ``"False"``olarak ayarlayarak aynı şey kolayca görüntülenebilir. Veri setimizde herhangi bir null değer bulunmadığından dropna parametresinin ayarlanması bir fark yaratmayacaktır. Ancak bu, boş değerlere sahip başka bir veri kümesinde faydalı olabilir; bu nedenle bunu aklınızda bulundurun.

In [31]:
#df['your_column'].value_counts(dropna=False)

## 8. value_counts() as dataframe
``"value_counts()"`` bir data frame i değil seriyi döndürür. Sayımlarınızın bir dataframe olarak dönmesini istiyorsanız, bunu ``".value_counts().to_frame()"``

In [32]:
# value_counts as dataframe
df['course_difficulty'].value_counts().to_frame()

count
course_difficulty       
Beginner             487
Intermediate         198
Mixed                187
Advanced              19